### PR TITLE
Refactors nanosuit code and fixes/rebalance stuff.

### PR DESCRIPTION
**Pull request by yoyobatty**
[Changelogs]: # (Your PR should contain a detailed changelog of notable changes, titled and categorized appropriately. This includes, new features, sprites, sounds, balance changes, admin tools, map edits, removals, big refactors, config changes, hosting changes and important fixes. An example changelog has been provided below for you to edit. If you need additional help, read https://github.com/tgstation/tgstation/wiki/Changelogs)

:cl: YoYoBatty
add: Proper regen system.
balance: Consumes energy faster, regens a bit slower, +regen delay changes.
fix: Fixed certain projectiles and attacks not draining energy properly.
fix: Throwing fast now only works in strength mode.
refactor: Completely cleaned up the energy recharge and discharge code.
/:cl:

[why]: # (Please add a short description [two lines down] of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.)

I've spent the past 3 days studying the original source code of Crysis only to scrap and rewrite it just to make this work. Please love it.

Basically, this will feel more like the nanosuit from the game Crysis which was the original goal here. A lot of the old hacky and gross code has been scrapped and completely redesigned. Testing has been acute and done according to a checklist this time. Cannot find a runtime for the life of me. Please code check this. I plan on making a video soon to highlight the new functionality.

### CODE REVIEW

**Comment by ghost** [comment]
+1

**Comment by Carbonhell** [comment]
+1

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 1]
```diff
@@ -277,21 +279,15 @@
...
-	else
-		cell = new(src)
-	cell.give(cell.maxcharge)
 	update_icon()
 	START_PROCESSING(SSobj, src)
+	cl_nn_energy = nn_energy
 
 /obj/item/clothing/suit/space/hardsuit/nano/Destroy()
 	STOP_PROCESSING(SSobj, src)
-	QDEL_NULL(cell)
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: why this?

**Comment by Carbonhell** [comment]
wait did you just cut out cell code and make a snowflake version of it?why?

**Comment by dreamcrab** [comment]
looks cool and if it aint broke +8

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 2]
```diff
@@ -302,61 +240,76 @@
...
-			if(cell.charge > 0)
-				cell.give(charge_rate) //this will get called after the bottom
-			else
-				sleep(40) //if we lose energy wait 4 seconds then recharge us
-				cell.give(charge_rate)
-		sleep(recharge_delay)//recharges us at variable rate
+	if(!U)
+		return
+	if(world.time > medical_cooldown && current_charges < max_charges)
+		current_charges = CLAMP((current_charges + 1), 0, max_charges)
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: Can't you just create a timer when a charge is used, that replenishes it after medical_cooldown time?

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 3]
```diff
@@ -302,61 +240,76 @@
...
+			temp_cooldown = world.time + restore_delay
+			detecting = TRUE
+		if(world.time > temp_cooldown)
+			if(!defrosted)
+				helmet.display_visor_message("Activating suit defrosting protocols.")
+				U.reagents.add_reagent("leporazine", 2)
+				defrosted = TRUE
+				temp_cooldown += 100
+	else
+		defrosted = FALSE
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: Setting those two vars to false every tick is pretty ugly. Can't you set them strictly when you need them?

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 4]
```diff
@@ -302,61 +240,76 @@
...
+		detecting = FALSE
+	if(!shutdown) //Is the suit working? Good, spawn a while loop
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: Is the code above supposed to run even if the suit is offline? Wat

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 5]
```diff
@@ -302,61 +240,76 @@
...
+		var/energy = cell.charge //store current energy here
+		if(mode == "cloak" && !U.Move()) //are we in cloak, not moving?
+			energy -= cl_energy * 0.1 //take away the cloak discharge rate at 1/10th since we're not moving
+		if((energy < cell.maxcharge) && mode != "cloak" && nn_block_recharge == 0) //if our energy is less than 100, we're not in cloak and don't have a recharge delay timer
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: var == 0 is ugly, just use !var reeeeeee

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 6]
```diff
@@ -302,61 +240,76 @@
...
+			var/energy2 = nn_regen //store our regen rate here
+			energy2+=energy //add our current energy to it
+			energy=min(cell.maxcharge,energy2) //our energy now equals the energy we had + 0.75 for everytime it iterates through, so it increases by 0.75 every tick until it goes to 100
+		if(nn_block_recharge > 0) //do we have a recharge delay set?
+			nn_block_recharge -= 1 //reduce it
+		if(msg_time_upper)
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: What are those three vars you are reducing? It's dangerous to assume this proc will run every x time and bad practice, it could go faster or slower if shit happens

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 7]
```diff
@@ -302,61 +240,76 @@
...
+	cell.charge = amount
+	if(delay > nn_block_recharge)
+		nn_block_recharge = delay
+	if(amount < cr_energy && !criticalpower) //energy is less than critical energy level(20) and not in crit power
+		helmet.display_visor_message("Energy Critical!") //now we are
+		criticalpower = TRUE
+	else if(amount > cr_energy) //did our energy go higher than the crit level
+		criticalpower = FALSE //turn it off
+	if(amount <= 0) //did we lose energy?
+		amount = 0 //set our energy to 0
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:291]: Why are you setting an arg's value? You probably meant to edit cell.charge here.

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 8]
```diff
@@ -464,6 +427,62 @@
 		var/datum/action/A = X
 		A.UpdateButtonIcon()
 
+/obj/item/clothing/suit/space/hardsuit/nano/emp_act(severity)
+	..()
+	set_nano_energy(max(0,cell.charge-(cell.charge/severity)),80)
+	if((mode == armor && cell.charge == 0) || (mode != armor))
+		if(prob(8/severity*1.5) && !shutdown)
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: severity can't be 0 right?

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 9]
```diff
@@ -464,6 +427,62 @@
...
+			emp_assault()
+	update_icon()
+
+/obj/item/clothing/suit/space/hardsuit/nano/proc/emp_assault()
+	if(!U.mind)
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: Why are you checking if the mind exists? You aren't even using it, you should rather check if U exists

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 10]
```diff
@@ -464,6 +427,62 @@
...
+		return //Not sure how this could happen.
+	U.Knockdown(300)
+	U.AdjustStun(300)
+	U.Jitter(120)
+	toggle_mode("none", TRUE)
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: Modes should be defines for style purposes and so that it is more easily maintainable

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 11]
```diff
@@ -619,7 +637,7 @@
 		var/obj/item/clothing/suit/space/hardsuit/nano/NS = wear_suit
 		if(statpanel("Crynet Nanosuit"))
 			stat("Crynet Protocols : Engaged")
-			stat("Energy Charge:", "[NS.cell.charge]%")
+			stat("Energy Charge:", "[round(NS.cell.charge)]%")
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: To keep this fully modular you could do NS.cell.charge*100/NS.cell.maxcharge (or max_charge idk) so that it returns a value between 0 and 100

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 12]
```diff
@@ -856,16 +870,14 @@
 			if(!ismob(O) || user.a_intent == INTENT_HARM) //melee attack
 				NS.kill_cloak()
 
-
 /obj/item/gun/attack(mob/M as mob, mob/living/carbon/user)
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:884]: Wait what is this for? I think this triggers if you bash someone with a gun, you already did this didn't you

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 4 above)
Comment (Carbonhell) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: Also outdated comments

**Review by Carbonhell — DISMISSED** [review]

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 5 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: meh

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 4 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: I gotta change

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 3 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: yup

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 2 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: that's what it does

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 8 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: severity should never be zero, but I'll put a check for it

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 9 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: probably a good idea, that code is super old

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 10 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: agreed

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 12 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:884]: afterattack is for firing and attack is for melee

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm (hunk 7 above)
Comment (yoyobatty) [reply hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:291]: woops

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 13]
```diff
@@ -213,150 +213,111 @@
...
+		criticalpower = TRUE
+	else if(amount > cr_energy) //did our energy go higher than the crit level
+		criticalpower = FALSE //turn it off
+	if(amount <= 0) //did we lose energy?
+		amount = 0 //set our energy to 0
+		if(mode == CLOAK) //are we in cloak?
+			nn_block_recharge = 30 //then wait 30 ticks to recharge again
+		if(mode != ARMOR) //we're not in cloak
+			toggle_mode(ARMOR, TRUE) //go into it, forced
+	return TRUE
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm]: Why return TRUE? Do you actually use this value? Doesn't seems so

File: hippiestation/code/modules/clothing/spacesuits/nanosuit.dm [hunk 14]
```diff
@@ -994,7 +1016,7 @@
 	to_chat(usr, "<b>Passive equipment</b>: Binoculars, night vision, anti-slips, shock and heat proof gloves, self refilling mini o2 tank, emergency medical systems and body temperature defroster.")
 	to_chat(usr, "<b>Active modes</b>: Armor, strength, speed and cloak.")
 	to_chat(usr, "<span class='notice'>Armor</span>: Resist damage that would normally kill or seriously injure you. Blocks all attacks at a cost of suit energy drain.")
-	to_chat(usr, "<span class='notice'>Cloak</span>: Become a ninja. Cloaking technology alters the outer layers to refract light through and around the suit, making the user appear almost completely invisible.")
+	to_chat(usr, "<span class='notice'>Cloak</span>: Become a ninja. Cloaking technology alters the outer layers to refract light through and around the suit, making the user appear almost completely invisible. Simple tasks such as attacking in any way, being hit or throwing objects cancels cloak.")
```
Comment (Carbonhell) [inline hippiestation/code/modules/clothing/spacesuits/nanosuit.dm:1020]: It doesn't just cancel the cloak, it empties your cell, should specify that

**Review by Carbonhell — CHANGES_REQUESTED** [review]

**Review by Carbonhell — APPROVED** [review]
